### PR TITLE
#312 DeleteView implementation and using it

### PR DESCRIPTION
**Pull request by vryazanov**
#312  
* DeleteView implemented
* delete view in table shape endpoint refactored

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#383](https://github.com/timelesslounge/timelessis/issues/383) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#383](https://github.com/timelesslounge/timelessis/issues/383) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/c05f67f4-9662-4aae-ba15-c1fe008bcf7e); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: tests/integration/it_table_shapes_test.py [hunk 1]
```diff
@@ -101,14 +102,9 @@ def test_edit(client):
...
-    client.post(flask.url_for("table_shape.create"), data={
-        "description": "It's new shape",
-        "picture": "http://...."
-    })
-    identifier = models.TableShape.query.first().id
-
-    result = client.post(flask.url_for('table_shape.delete', id=identifier))
-    assert result.status_code == HTTPStatus.FOUND
-    assert models.TableShape.query.filter_by(id=identifier).count() == 0
+    table_shape = factories.TableShapeFactory()
```
Comment (amihaiemil) [inline tests/integration/it_table_shapes_test.py]: @vryazanov Shouldn't there be one method call here, like ``.getTableShape()``? As in, the Factory creates the table_shape, right?

File: tests/integration/it_table_shapes_test.py [hunk 2]
```diff
@@ -101,14 +102,9 @@ def test_edit(client):
...
+    response = client.post(
+        flask.url_for('table_shape.delete', id=table_shape.id))
```
Comment (amihaiemil) [inline tests/integration/it_table_shapes_test.py]: @vryazanov before deleting it, shouldn't you call ``/create`` to post it on the server? Otherwise, what is actually deleted?

File: timeless/restaurants/table_shapes/views.py [hunk 3]
```diff
@@ -73,10 +74,14 @@ def edit(id):
...
+    """
+    @todo #173:30min Refactor all deleting views to use `views.DeleteView`
+     and the method they were registered to bluprints. See `.register` method
+     in `GenericView`, use it. Also uncomment all tests related to these views.
+    """
+    model = models.TableShape
+    success_view_name = "table_shape.list"
+
+
+Delete.register(bp, "/<int:id>/delete")
```
Comment (amihaiemil) [inline timeless/restaurants/table_shapes/views.py]: @vryazanov shouldn't it be ``/delete/<int:id>``?

File: timeless/restaurants/table_shapes/views.py [hunk 4]
```diff
@@ -73,10 +74,14 @@ def edit(id):
...
-@bp.route("/delete/<int:id>", methods=["POST"])
-def delete(id):
-    """ Delete table shape with id """
-    table_shape = models.TableShape.query.get(id)
-    DB.session.delete(table_shape)
-    DB.session.commit()
-    return redirect(url_for("table_shape.list"))
+class Delete(views.DeleteView):
+    """
+    @todo #173:30min Refactor all deleting views to use `views.DeleteView`
```
Comment (amihaiemil) [inline timeless/restaurants/table_shapes/views.py]: @vryazanov ``#312`` here, not ``#173``, since that is the originating ticket :)

File: timeless/views.py [hunk 5]
```diff
@@ -189,14 +188,27 @@ def get_default_context(self):
...
     model = None
+    object_url_lookup = "id"
+
+    def get_object(self):
+        """ Takes object based on id provided in URL """
+        object_id = self.kwargs.get(self.object_url_lookup)
+        return self.model.query.get(object_id)
+
+
+class SuccessRedirectMixin:
```
Comment (amihaiemil) [inline timeless/views.py]: @vryazanov This SuccessRedirectMixin was clearly not part of the task, why did you add it? What about some tests for it?

File: timeless/views.py [hunk 6]
```diff
@@ -266,9 +262,15 @@ class UpdateView(GenericView):
     """ Base view for updating objects"""
 
 
-class DeleteView(GenericView):
+class DeleteView(SuccessRedirectMixin, SingleObjectMixin, GenericView):
     """ Base view for deleting objects """
```
Comment (amihaiemil) [inline timeless/views.py]: @vryazanov Can you add more details to this docstring? Maybe point the reader to a place where it is used, to see it in action. I hate that apparently people don't write docstrings/javadocs in their pythin code.

File: timeless/views.py [hunk 7]
```diff
@@ -189,14 +188,27 @@ def get_default_context(self):
...
-    def get_object(self, id=None):
-        """ Method fetch object from given model by id """
-        assert self.model, "Model is not provided"
-        return self.model.query.get(id)
+    def get_success_url_redirect(self):
+        """ Reverse URL based on view name """
+        if not self.success_view_name:
+            raise NotImplementedError(f"{self.__class__.__name__} must define "
+                                      f"'success_view_name' attribute")
+        return url_for(self.success_view_name)
```
Comment (amihaiemil) [inline timeless/views.py]: @vryazanov I think this mixin should actually perform the redirect, not just read the URL. If it just reads the url it's not of much help, is it?

File: tests/integration/it_table_shapes_test.py (hunk 2 above)
Comment (vryazanov) [reply tests/integration/it_table_shapes_test.py]: @amihaiemil  `factories.TableShapeFactory()` creates an instance in DB which we are going to delete

File: timeless/views.py (hunk 5 above)
Comment (vryazanov) [reply timeless/views.py]: @amihaiemil this logic already was in `CreateView` and it's needed to me in `DeleteView`, so that's why I moved it to mixin class.

File: timeless/views.py (hunk 7 above)
Comment (vryazanov) [reply timeless/views.py]: @amihaiemil hm, sounds reasonable, mb here should be a method which redirects user. but as I said before it's already implemented logic so some classes already relу on current behavior. I just moved it to a separate base class. I think we can create a bug to polish the design of generic views.

File: tests/integration/it_table_shapes_test.py (hunk 1 above)
Comment (vryazanov) [reply tests/integration/it_table_shapes_test.py]: @amihaiemil actually factories work a bit different. You can have a look at example from library we are using: https://factoryboy.readthedocs.io/en/latest/#using-factories
I can change the code to `factories.TableShapeFactory.create()` to make it more understandable. wdyt?

File: timeless/restaurants/table_shapes/views.py (hunk 4 above)
Comment (vryazanov) [reply timeless/restaurants/table_shapes/views.py]: @amihaiemil you are absolutely right, thanks!

File: timeless/restaurants/table_shapes/views.py (hunk 3 above)
Comment (vryazanov) [reply timeless/restaurants/table_shapes/views.py]: @amihaiemil yes, it should :+1: my bad

**Comment by vryazanov** [comment]
@amihaiemil I've answered to your questions and fixed the code. Please have a look and let me know  if I forgot something.

**Comment by amihaiemil** [comment]
@vryazanov looks good, thanks! :)

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@vryazanov there is a conflict, please fix

**Comment by vryazanov** [comment]
@emilianodellacasa done, pls have a look

**Comment by emilianodellacasa** [comment]
@vryazanov Sorry there is another conflict, cannot merge

**Comment by vryazanov** [comment]
@emilianodellacasa done, pls have a look

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#383](https://github.com/timelesslounge/timelessis/issues/383) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)
